src/SFML/Window/WindowImpl.cpp

```diff
@@ -175,30 +176,50 @@ void WindowImpl::setMaximumSize(const std::optional<Vector2u>& maximumSize)
 
 
 ////////////////////////////////////////////////////////////
-Event WindowImpl::popEvent(bool block)
+Event WindowImpl::popEventBlocking(Time timeout)
 {
+    const auto startTime = std::chrono::steady_clock::now();
+
+    const auto timedOut = [&]
+    { return timeout != Time::Zero && (std::chrono::steady_clock::now() - startTime) < timeout.toDuration(); };
+
     // If the event queue is empty, let's first check if new events are available from the OS
     if (m_events.empty())
     {
         // Get events from the system
-        processJoystickEvents();
-        processSensorEvents();
-        processEvents();
+        populateEventQueue();
+    }
 
-        // In blocking mode, we must process events until one is triggered
-        if (block)
-        {
-            // Here we use a manual wait loop instead of the optimized
-            // wait-event provided by the OS, so that we don't skip joystick
-            // events (which require polling)
-            while (m_events.empty())
-            {
-                sleep(milliseconds(10));
-                processJoystickEvents();
-                processSensorEvents();
-                processEvents();
-            }
-        }
+    // Here we use a manual wait loop instead of the optimized
+    // wait-event provided by the OS, so that we don't skip joystick
+    // events (which require polling)
+    while (m_events.empty() && !timedOut())
+    {
+        sleep(milliseconds(10));
+        populateEventQueue();
+    }
+
+    Event event;
+
+    // Pop the first event of the queue, if it is not empty
+    if (!m_events.empty())
+    {
+        event = m_events.front();
+        m_events.pop();
+    }
+
+    return event;
```

Comment:
Maybe this shared code piece can be moved out to e.g. `popEvent`

Reply:
Will extract